File: backend/api/schemas/composition.py
import re
from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

_HEX_COLOR = re.compile(r"^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{6})$")


def _validate_hex_color(value: str) -> str:
    if not _HEX_COLOR.match(value):
        raise ValueError("color must be a hex value like #FFF or #FFFFFF")
    return value


class OverlayAnchor(str, Enum):
    top_left = "top-left"
    top_center = "top-center"
    top_right = "top-right"
    center_left = "center-left"
    center = "center"
    center_right = "center-right"
    bottom_left = "bottom-left"
    bottom_center = "bottom-center"
    bottom_right = "bottom-right"


class OverlayPlacement(BaseModel):
    """Absolute placement for a fixed-position overlay.

    `offset_x`/`offset_y` are insets from the anchored edge. `max_width`/
    `max_height` bound the available area; when omitted, the area is the image
    minus the offsets on that side.
    """

    anchor: OverlayAnchor = OverlayAnchor.bottom_right
    offset_x: int = Field(default=24, ge=0)
    offset_y: int = Field(default=24, ge=0)
    max_width: int | None = Field(default=None, gt=0)
    max_height: int | None = Field(default=None, gt=0)


class OverlayStyle(BaseModel):
    font_family: str = Field(default="default", min_length=1, max_length=80)
    font_size: int = Field(default=48, gt=0, le=2000)
    color: str = "#FFFFFF"
    opacity: float = Field(default=1.0, ge=0.0, le=1.0)
    background_color: str | None = None
    stroke_color: str | None = None
    stroke_width: int = Field(default=0, ge=0, le=50)

    @field_validator("color")
    @classmethod
    def validate_color(cls, value: str) -> str:
        return _validate_hex_color(value)

    @field_validator("background_color", "stroke_color")
    @classmethod
    def validate_optional_color(cls, value: str | None) -> str | None:
        return _validate_hex_color(value) if value is not None else None


class CompositionRequest(BaseModel):
    sku: str = Field(min_length=1, max_length=200)
    placement: OverlayPlacement = Field(default_factory=OverlayPlacement)
    style: OverlayStyle = Field(default_factory=OverlayStyle)


class FitResult(BaseModel):
    fits: bool
    rendered_width: int
    rendered_height: int
    available_width: int
    available_height: int
    reason: str | None = None


class CompositionVersionResponse(BaseModel):
    id: UUID
    overlay_id: UUID
    generation_job_id: UUID
    version: int
    status: str
    sku: str
    placement: OverlayPlacement
    style: OverlayStyle
    font_version: str
    rendered_key: str | None
    rendered_checksum: str | None
    fit_result: FitResult
    created_at: datetime

    @classmethod
    def from_model(cls, version, generation_job_id: UUID) -> "CompositionVersionResponse":
        return cls(
            id=version.id,
            overlay_id=version.overlay_id,
            generation_job_id=generation_job_id,
            version=version.version,
            status=version.status,
            sku=version.sku_normalized,
            placement=OverlayPlacement(**version.placement),
            style=OverlayStyle(**version.style),
            font_version=version.font_version,
            rendered_key=version.rendered_key,
            rendered_checksum=version.rendered_checksum,
            fit_result=FitResult(**version.fit_result),
            created_at=version.created_at,
        )


class CompositionSummaryResponse(BaseModel):
    overlay_id: UUID
    generation_job_id: UUID
    base_image_key: str
    latest_version: CompositionVersionResponse | None

File: frontend/lib/api/composition.ts
import { ApiError, apiFetch } from '@/lib/api';

export type OverlayAnchor =
  | 'top-left'
  | 'top-center'
  | 'top-right'
  | 'center-left'
  | 'center'
  | 'center-right'
  | 'bottom-left'
  | 'bottom-center'
  | 'bottom-right';

export interface OverlayPlacement {
  anchor: OverlayAnchor;
  offset_x: number;
  offset_y: number;
  max_width?: number | null;
  max_height?: number | null;
}

export interface OverlayStyle {
  font_family: string;
  font_size: number;
  color: string;
  opacity: number;
  background_color?: string | null;
  stroke_color?: string | null;
  stroke_width: number;
}

export interface CompositionRequest {
  sku: string;
  placement?: OverlayPlacement;
  style?: OverlayStyle;
}

export interface FitResult {
  fits: boolean;
  rendered_width: number;
  rendered_height: number;
  available_width: number;
  available_height: number;
  reason: string | null;
}

export interface CompositionVersion {
  id: string;
  overlay_id: string;
  generation_job_id: string;
  version: number;
  status: string;
  sku: string;
  placement: OverlayPlacement;
  style: OverlayStyle;
  font_version: string;
  rendered_key: string | null;
  rendered_checksum: string | null;
  fit_result: FitResult;
  created_at: string;
}

export async function composeSkuOverlay(
  jobId: string,
  body: CompositionRequest,
): Promise<CompositionVersion> {
  return apiFetch(`/api/generation-jobs/${jobId}/composition`, {
    method: 'POST',
    body: JSON.stringify(body),
  }) as Promise<CompositionVersion>;
}

export async function listCompositionVersions(jobId: string): Promise<CompositionVersion[]> {
  try {
    return (await apiFetch(
      `/api/generation-jobs/${jobId}/composition/versions`,
    )) as CompositionVersion[];
  } catch (error) {
    if (error instanceof ApiError && error.status === 404) {
      return [];
    }
    throw error;
  }
}
